keanu-project/src/main/java/io/improbable/keanu/network/BayesianNetwork.java

```diff
@@ -214,15 +222,49 @@ public void incrementIndentation() {
         indentation++;
     }
 
-    public void save(NetworkWriter networkWriter) throws IOException {
+    public void save(NetworkWriter networkWriter) {
         for (Vertex vertex : TopologicalSort.sort(vertices)) {
             vertex.save(networkWriter);
         }
     }
 
-    public void saveValues(NetworkWriter networkWriter) throws IOException {
+    public void saveValues(NetworkWriter networkWriter) {
         for (Vertex vertex : vertices) {
             vertex.saveValue(networkWriter);
         }
     }
+
+    /**
+     * Method for traversing a graph and returning a subgraph of vertices within the given degree of the specified vertex.
+     *
+     * @param vertex vertex that the subgraph will be centered around
+     * @param degree degree of connections from the vertex to be included in the subgraph
+     * @return a set of vertices within the specified degree from the given vertex
+     */
+    public Set<Vertex> getSubgraph(Vertex vertex, int degree) {
+
+        Set<Vertex> subgraphVertices = new HashSet<>();
+        Set<Vertex> verticesToProcessNow = new HashSet<>();
```

Comment:
For `subgraphVertices` I believe hashset is better, as I'm doing finds against it (`if (!subgraphVertices.contains(connectedVertex))`). For `verticesToProcessNow` - if i have `verticesToProcessNext` as a hashset I would have to convert that to a list before iterating over it. Not sure about the complexity of this conversion, but I'm not convinced it would be faster. What did you have in mind?

Reply:
I think if we just re-ordered how things are done - we could leave the "toProcess" things as Lists and keep the subgraph vertex as a set.  If we stick things in the subgraphVertices set before we process them we can simply check against that?